packages/base-controller/src/ControllerMessenger.ts

```diff
@@ -41,37 +59,41 @@ export type SelectorEventHandler<SelectorReturnValue> = (
 
 export type ActionConstraint = {
   type: string;
-  handler: (...args: any) => unknown;
+  handler: (...args: never[]) => unknown;
```

Comment:
This change seems to be responsible for breaking the types in the rate limit controller. If we remove this change, the rate limit controller changes can be removed as well, and this becomes a non-breaking PR.

Reply:
Hmm.... it seems to be non-breaking if we make it `never` though, rather than `never[]`. At the moment it's typed as `any`; even changing it to `any[]` introduces type errors.

Reply:
Addressed here: d474b253032f098605f32321246a5e736605dd9c. I think this makes `ActionConstraint` non-breaking?

I tentatively typed `ReturnType<RateLimitedApi['method']>` as `boolean` based on the specified [jsdoc return type](https://github.com/MetaMask/core/pull/1890/commits/d474b253032f098605f32321246a5e736605dd9c#diff-1eb134c9c5a9dd0a4e4838f719ba67723f61fdacd8bfc1a9acef96366b7578fbR149), but replacing the `boolean` with `unknown` also works.

Reply:
Hmm, it still seems like there are breaking changes here to the base controller package. I tried undoing the `rate-limit-controller` changes and there are two type errors:

```
packages/rate-limit-controller/src/RateLimitController.ts:57:3 - error TS2344: Type 'RateLimitControllerActions<RateLimitedApis>' does not satisfy the constraint 'ActionConstraint'.
  Type 'CallApi<RateLimitedApis>' is not assignable to type 'ActionConstraint'.
    Types of property 'handler' are incompatible.                                                                                                                                          
      Type '<ApiType extends keyof RateLimitedApis>(origin: string, type: ApiType, ...args: Parameters<RateLimitedApis[ApiType]["method"]>) => Promise<...>' is not assignable to type '(..
.args: never[]) => unknown'.                                                                                                                                                                       Types of parameters 'origin' and 'args' are incompatible.
          Type 'never[]' is not assignable to type '[origin: string, type: never, ...args: Parameters<RateLimitedApis[never]["method"]>]'.
            Target requires 3 element(s) but source may have fewer.                                                                                                                        

57   RateLimitControllerActions<RateLimitedApis>,                                                                                                                                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~   

packages/rate-limit-controller/src/RateLimitController.ts:75:3 - error TS2344: Type 'RateLimitMessenger<RateLimitedApis>' does not satisfy the constraint 'RestrictedControllerMessenger<"R
ateLimitController", ActionConstraint | ControllerActions<"RateLimitController", RateLimitState<RateLimitedApis>>, EventConstraint | ControllerEvents<...>, string, string>'.
  Types of property 'registerActionHandler' are incompatible.
    Type '<ActionType extends "RateLimitController:getState" | "RateLimitController:call">(action: ActionType, handler: ActionHandler<RateLimitControllerActions<RateLimitedApis>, ActionTy
pe>) => void' is not assignable to type '<ActionType extends string>(action: ActionType, handler: ActionHandler<ActionConstraint | ControllerActions<"RateLimitController", RateLimitState<
RateLimitedApis>>, ActionType>) => void'.
      Types of parameters 'handler' and 'handler' are incompatible.      
        Types of parameters 'args' and 'args' are incompatible.
          Type '[] | [origin: string, type: keyof RateLimitedApis, ...args: Parameters<RateLimitedApis[keyof RateLimitedApis]["method"]>]' is not assignable to type 'never[] | []'.
            Type '[origin: string, type: keyof RateLimitedApis, ...args: Parameters<RateLimitedApis[keyof RateLimitedApis]["method"]>]' is not assignable to type 'never[] | []'.
              Type '[origin: string, type: keyof RateLimitedApis, ...args: Parameters<RateLimitedApis[keyof RateLimitedApis]["method"]>]' is not assignable to type 'never[]'.
                Type 'string | keyof RateLimitedApis | Parameters<RateLimitedApis[keyof RateLimitedApis]["method"]>[number]' is not assignable to type 'never'.
                  Type 'string' is not assignable to type 'never'.

75   RateLimitMessenger<RateLimitedApis>

```

Reply:
This is caused by the removal of `any` in `ActionContraint['handler']` itself and in the generic params for ControllerMessenger. 

The change that resolves this issue is at [lines 52-54 of rate=limit-controller](https://github.com/MetaMask/core/pull/1890/files/1fa24a826e4bec9c1e9e01aea8dbb3105438528a#diff-1eb134c9c5a9dd0a4e4838f719ba67723f61fdacd8bfc1a9acef96366b7578fbR52-R54).

In all of these cases, no actual constraints are being added, since `ActionConstraint['handler']` is the supertype of all functions, but since previously suppressed errors are popping up, I added a **BREAKING** notice in the changelog for `ActionConstraint` and `EventConstraint`.

Reply:
Ah, so it sounds like the reason why this breaks `rate-limit-controller` is that the types there were incorrect. Is this true?

Reply:
Yes the constraint in `RestrictedControllerMessenger` enforcing that `CallApi extends ActionConstraint` was already in place -- it was just being suppressed by the `any` usage (the definition of `this.call()` and `CallApi` both derive from `RateLimitedApi['method']`, which was previously typed with `any`).